qoffer-infrastructure/src/main/groovy/life/qbic/portal/qoffer2/offers/OfferDbConnector.groovy

```diff
@@ -17,19 +22,95 @@ import org.apache.logging.log4j.Logger
+            connection.setAutoCommit(false)
+
+            connection.withCloseable {it ->
+                try {
+                    int projectManagerId = offerToCustomerGateway.getPersonId(offer.projectManager)
+                    int customerId = offerToCustomerGateway.getPersonId(offer.customer)
+                    int affiliationId = offerToCustomerGateway.getAffiliationId(offer.selectedCustomerAffiliation)
+
+                    int offerId = storeOffer(offer, projectManagerId, customerId, affiliationId)
+
+                    offerToProductGateway.createOfferItems(offer.items, offerId)
+                    connection.commit()
+                } catch (Exception e) {
+                    log.error(e.message)
+                    log.error(e.stackTrace.join("\n"))
+                    connection.rollback()
+                    connection.close()
+                    throw new DatabaseQueryException("Could not store offer.")
+                }
+            }
+        } catch (DatabaseQueryException ignored) {
+            throw new DatabaseQueryException("The offer could not be stored: ${offer.toString()}")
+        } catch (Exception e) {
+            log.error(e)
+            log.error(e.stackTrace.join("\n"))
+            throw new DatabaseQueryException("The offer could not be stored: ${offer.toString()}")
+        }
+    }
+
+    /**
+     * The method stores the offer in the QBiC database
+     *
+     * @param offer with the information of the offer to be stored
+     * @return the id of the stored offer in the database
+     */
+    private int storeOffer(Offer offer, int projectManagerId, int customerId, int affiliationId){
+        String sqlValues = "VALUE(?,?,?,?,?,?,?,?)"
+        String queryTemplate = OFFER_INSERT_QUERY + " " + sqlValues
+
+        try{
+            List<Integer> generatedKeys = []
+            Connection connection = connectionProvider.connect()
+
+            connection.withCloseable {
+                PreparedStatement preparedStatement = it.prepareStatement(queryTemplate, Statement.RETURN_GENERATED_KEYS)
+                preparedStatement.setDate(1, new Date(offer.modificationDate.time))
+                preparedStatement.setDate(2, new Date(offer.expirationDate.time))
+                preparedStatement.setInt(3,customerId)
+                preparedStatement.setInt(4,projectManagerId)
+                preparedStatement.setString(5,offer.projectTitle)
+                preparedStatement.setString(6,offer.projectDescription)
+                preparedStatement.setDouble(7,offer.totalPrice)
+                preparedStatement.setInt(8,affiliationId)
+
+                preparedStatement.execute()
+
+                def keys = preparedStatement.getGeneratedKeys()
+                while (keys.next()){
+                    generatedKeys.add(keys.getInt(1))
+                }
+
+                return generatedKeys[0]
+            }
+        }catch(Exception exception){
```

Comment:
Suggestion: Make the try catch around the method call, not inside the method -> much cleaner and more explicit, where the exception is catched.

Reply:
dc8406c